File: src/main/Fsharp/eyefuck.fs
(*
i actully love f# but i only learned it for 2 years
*)

open System
open System.IO
open System.Text.RegularExpressions

// ANSI colors
let RESET = "\u001b[0m"
let RED = "\u001b[31m"
let GREEN = "\u001b[32m"
let YELLOW = "\u001b[33m"
let BLUE = "\u001b[34m"
let CYAN = "\u001b[36m"
let WHITE = "\u001b[97m"

let EYF_V = 1.2
let TAPE_SIZE = 300000

// ---------------------------
// Main function
// ---------------------------
[<EntryPoint>]
let main argv =
    if argv.Length < 1 then
        printfn "%sUsage:%s eyefuck <command> [file.eyf]" RED RESET
        1
    else
        let mode = argv.[0]

        match mode with
        | "run" ->
            if argv.Length < 2 then
                printfn "%sPlease specify a file to run.%s" RED RESET
                1
            else
                let file = argv.[1]
                try
                    let code = File.ReadAllText(file)
                    runInterpreter code
                    0
                with
                | e ->
                    printfn "Error reading file: %s" e.Message
                    1

        | "-i" | "--i" | "i" ->
            startREPL()
            0

        | "help" | "-help" | "-h" | "--h" | "--help" ->
            printfn "%sEyefuck HELP:%s" CYAN RESET
            printfn "%s  eyefuck run <file.eyf>%s  -> %sexecute the Eyefuck file%s" YELLOW RESET GREEN RESET
            printfn "%s  eyefuck -i%s             -> %sinteractive REPL mode%s" YELLOW RESET GREEN RESET
            printfn "%s  eyefuck about%s          -> %sinformation about this interpreter%s" YELLOW RESET GREEN RESET
            0

        | "about" ->
            printfn "%sEyefuck DEV 2025%s" CYAN RESET
            printfn "%sMIT license%s see LICENSE for more information" GREEN RESET
            printfn "Please help me motive by giving the repo a star"
            printfn "%sgithub:%s github.com/bandikaaking" BLUE RESET
            printfn "crafted with %s<3%s by %s@Bandikaaking%s" RED RESET YELLOW RESET
            0

        | "version" | "--v" | "--version" | "-v" | "v" | "-version" ->
            printfn "Current eyefuck version: %f" EYF_V
            0

        | "ov" | "-ov" | "--ov" ->
            printfn "Other Eyefuck versions: "
            printfn "0.10: Started / added 2 instructions"
            printfn "0.11-0.43: Fixed many bugs, and edded 5 more instructions"
            printfn "1.0: Added syntax highliting"
            printfn "1.1: Fixed bugs"
            printfn "added more eyefuck modes / rewrited README.md"
            0

        | _ ->
            printfn "%sUnknown mode:%s %s" RED RESET mode
            1

// ---------------------------
// Interactive REPL
// ---------------------------
let startREPL () =
    printfn "%sEyefuck DEV 2025 - REPL%s" CYAN RESET
    printfn "Type commands below, empty line to execute, Ctrl+C to exit"
    
    let rec readLines lines =
        printf "$ "
        let line = Console.ReadLine()
        
        if String.IsNullOrWhiteSpace(line) then
            runInterpreter (String.Join("\n", List.rev lines))
            readLines []
        else
            readLines (line :: lines)
    
    try
        readLines []
    with
    | _ -> ()

// ---------------------------
// Eyefuck Interpreter
// ---------------------------
let runInterpreter (code: string) =
    let tape = Array.zeroCreate<byte> TAPE_SIZE
    let mutable ptr = 0
    let lines = code.Split('\n')
    let mutable loopStack = []
    let mutable i = 0

    while i < lines.Length do
        let line = lines.[i].Trim()
        
        // remove comments after #
        let line = 
            match line.IndexOf('#') with
            | -1 -> line
            | idx -> line.Substring(0, idx).Trim()
        
        if not (String.IsNullOrEmpty(line)) then
            match line with
            | "^" -> // increment cell
                tape.[ptr] <- tape.[ptr] + 1uy
            | "v" -> // decrement cell
                tape.[ptr] <- tape.[ptr] - 1uy
            | ">" -> // move pointer right
                ptr <- (ptr + 1) % TAPE_SIZE
            | "<" -> // move pointer left
                ptr <- if ptr = 0 then TAPE_SIZE - 1 else ptr - 1
            | line when line.StartsWith("bin") -> // set cell from binary
                let binStr = line.Substring(3).Trim()
                try
                    tape.[ptr] <- Convert.ToByte(binStr, 2)
                with
                | _ -> 
                    printfn "Invalid binary format"
                    exit 1
            | line when line.StartsWith("col") -> // set text color from HEX
                let m = Regex.Match(line, @"\[([0-9A-Fa-f]+)\]")
                if m.Success then
                    let hex = m.Groups.[1].Value
                    try
                        let colorInt = Convert.ToInt32(hex, 16)
                        let r = (colorInt >>> 16) &&& 0xFF
                        let g = (colorInt >>> 8) &&& 0xFF
                        let b = colorInt &&& 0xFF
                        printf "\u001b[38;2;%d;%d;%dm" r g b
                    with
                    | _ -> 
                        printfn "Invalid HEX color"
                        exit 1
            | line when line.StartsWith("load[") -> // load file
                let m = Regex.Match(line, @"\[([^\]]+)\]")
                if m.Success then
                    let filename = m.Groups.[1].Value
                    try
                        File.ReadAllBytes(filename) |> ignore
                        tape.[ptr] <- 0uy
                    with
                    | _ -> 
                        printfn "Error loading file"
                        exit 1
            | "," -> // read single byte input
                let input = Console.Read()
                if input <> -1 then
                    tape.[ptr] <- byte input
            | "." -> // print cell as char
                printf "%c" (char tape.[ptr])
            | "loop[" -> // start loop
                loopStack <- i :: loopStack
            | "]" -> // end loop
                if tape.[ptr] <> 0uy then
                    match loopStack with
                    | head :: tail ->
                        i <- head
                        loopStack <- tail
                    | [] ->
                        printfn "Unmatched ]"
                        exit 1
                else
                    loopStack <- List.tail loopStack
            | _ ->
                printfn "%serror caught while parsing" RED
                printfn "%sat line: %s" RED line
                exit 1
        
        i <- i + 1
    
    printfn ""

File: src/main/C/eyefuck.c
// c version,
/* i did this... with only 2 weeks?!!?? wow i am lazy as HELL, i wroted the eyefuck.lua / eyefuck.rb  under 2 hours. Sooo i guess i am lazy*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

// ANSI colors
#define RESET "\033[0m"
#define RED "\033[31m"
#define GREEN "\033[32m"
#define YELLOW "\033[33m"
#define BLUE "\033[34m"
#define CYAN "\033[36m"
#define WHITE "\033[97m"

#define EYF_V 1.2
#define TAPE_SIZE 300000

// Function prototypes
void run_interpreter(const char *code);
void start_repl();

// ---------------------------
// Main function
// ---------------------------
int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf(RED "Usage:" RESET " eyefuck <command> [file.eyf]\n");
        return 1;
    }

    char *mode = argv[1];

    if (strcmp(mode, "run") == 0) {
        if (argc < 3) {
            printf(RED "Please specify a file to run." RESET "\n");
            return 1;
        }
        char *file = argv[2];
        FILE *fp = fopen(file, "r");
        if (!fp) {
            perror("Error opening file");
            return 1;
        }
        fseek(fp, 0, SEEK_END);
        long file_size = ftell(fp);
        fseek(fp, 0, SEEK_SET);
        char *code = malloc(file_size + 1);
        fread(code, 1, file_size, fp);
        code[file_size] = '\0';
        fclose(fp);
        run_interpreter(code);
        free(code);
    }
    else if (strcmp(mode, "-i") == 0 || strcmp(mode, "--i") == 0 || strcmp(mode, "i") == 0) {
        start_repl();
    }
    else if (strcmp(mode, "help") == 0 || strcmp(mode, "-help") == 0 || 
             strcmp(mode, "-h") == 0 || strcmp(mode, "--h") == 0 || strcmp(mode, "--help") == 0) {
        printf(CYAN "Eyefuck HELP:" RESET "\n");
        printf(YELLOW "  eyefuck run <file.eyf>" RESET "  -> " GREEN "execute the Eyefuck file" RESET "\n");
        printf(YELLOW "  eyefuck -i" RESET "             -> " GREEN "interactive REPL mode" RESET "\n");
        printf(YELLOW "  eyefuck about" RESET "          -> " GREEN "information about this interpreter" RESET "\n");
    }
    else if (strcmp(mode, "about") == 0) {
        printf(CYAN "Eyefuck DEV 2025" RESET "\n");
        printf(GREEN "MIT license" RESET " see LICENSE for more information\n");
        printf("Please help me motive by giving the repo a star\n");
        printf(BLUE "github:" RESET " github.com/bandikaaking\n");
        printf("crafted with " RED "<3" RESET " by " YELLOW "@Bandikaaking" RESET "\n");
    }
    else if (strcmp(mode, "version") == 0 || strcmp(mode, "--v") == 0 || 
             strcmp(mode, "--version") == 0 || strcmp(mode, "-v") == 0 || 
             strcmp(mode, "v") == 0 || strcmp(mode, "-version") == 0) {
        printf("Current eyefuck version: %.1f\n", EYF_V);
    }
    else if (strcmp(mode, "ov") == 0 || strcmp(mode, "-ov") == 0 || strcmp(mode, "--ov") == 0) {
        printf("Other Eyefuck versions: \n");
        printf("0.10: Started / added 2 instructions\n");
        printf("0.11-0.43: Fixed many bugs, and edded 5 more instructions\n");
        printf("1.0: Added syntax highliting\n");
        printf("1.1: Fixed bugs\n");
        printf("added more eyefuck modes / rewrited README.md\n");
    }
    else {
        printf(RED "Unknown mode:" RESET " %s\n", mode);
        return 1;
    }

    return 0;
}

// ---------------------------
// Interactive REPL
// ---------------------------
void start_repl() {
    printf(CYAN "Eyefuck DEV 2025 - REPL" RESET "\n");
    printf("Type commands below, empty line to execute, Ctrl+C to exit\n");
    
    char *code_lines = malloc(4096);
    code_lines[0] = '\0';
    size_t total_size = 4096;
    size_t current_len = 0;
    
    while (1) {
        printf("$ ");
        char line[256];
        if (!fgets(line, sizeof(line), stdin)) {
            break;
        }
        
        // Remove newline
        line[strcspn(line, "\n")] = '\0';
        
        if (strlen(line) == 0) {
            run_interpreter(code_lines);
            code_lines[0] = '\0';
            current_len = 0;
            continue;
        }
        
        // Resize if needed
        size_t line_len = strlen(line);
        if (current_len + line_len + 2 > total_size) {
            total_size *= 2;
            code_lines = realloc(code_lines, total_size);
        }
        
        strcat(code_lines, line);
        strcat(code_lines, "\n");
        current_len += line_len + 1;
    }
    
    free(code_lines);
}

// ---------------------------
// Eyefuck Interpreter
// ---------------------------
void run_interpreter(const char *code) {
    unsigned char *tape = calloc(TAPE_SIZE, sizeof(unsigned char));
    int ptr = 0;
    
    char *code_copy = strdup(code);
    char *lines[10000];
    int line_count = 0;
    
    // Split into lines
    char *line = strtok(code_copy, "\n");
    while (line && line_count < 10000) {
        lines[line_count++] = line;
        line = strtok(NULL, "\n");
    }
    
    int loop_stack[1000];
    int loop_stack_ptr = 0;
    
    for (int i = 0; i < line_count; i++) {
        char current_line[256];
        strncpy(current_line, lines[i], sizeof(current_line));
        current_line[sizeof(current_line) - 1] = '\0';
        
        // Remove comments after #
        char *comment_pos = strchr(current_line, '#');
        if (comment_pos) {
            *comment_pos = '\0';
        }
        
        // Trim whitespace
        char *trimmed = current_line;
        while (isspace(*trimmed)) trimmed++;
        char *end = trimmed + strlen(trimmed) - 1;
        while (end > trimmed && isspace(*end)) end--;
        *(end + 1) = '\0';
        
        if (strlen(trimmed) == 0) {
            continue;
        }

        if (strcmp(trimmed, "^") == 0) {
            // increment cell
            tape[ptr]++;
        }
        else if (strcmp(trimmed, "v") == 0) {
            // decrement cell
            tape[ptr]--;
        }
        else if (strcmp(trimmed, ">") == 0) {
            // move pointer right
            ptr = (ptr + 1) % TAPE_SIZE;
        }
        else if (strcmp(trimmed, "<") == 0) {
            // move pointer left
            ptr = (ptr == 0) ? TAPE_SIZE - 1 : ptr - 1;
        }
        else if (strncmp(trimmed, "bin", 3) == 0) {
            // set cell from binary
            char *bin_str = trimmed + 3;
            while (isspace(*bin_str)) bin_str++;
            tape[ptr] = (unsigned char)strtol(bin_str, NULL, 2);
        }
        else if (strncmp(trimmed, "col", 3) == 0) {
            // set text color from HEX
            char *start = strchr(trimmed, '[');
            char *end = strchr(trimmed, ']');
            if (start && end && end > start + 1) {
                *end = '\0';
                long color_int = strtol(start + 1, NULL, 16);
                int r = (color_int >> 16) & 0xFF;
                int g = (color_int >> 8) & 0xFF;
                int b = color_int & 0xFF;
                printf("\033[38;2;%d;%d;%dm", r, g, b);
            }
        }
        else if (strncmp(trimmed, "load[", 5) == 0) {
            // load file
            char *start = strchr(trimmed, '[');
            char *end = strchr(trimmed, ']');
            if (start && end && end > start + 1) {
                *end = '\0';
                FILE *fp = fopen(start + 1, "r");
                if (fp) {
                    fclose(fp);
                    tape[ptr] = 0;
                }
            }
        }
        else if (strcmp(trimmed, ",") == 0) {
            // read single byte input
            tape[ptr] = getchar();
        }
        else if (strcmp(trimmed, ".") == 0) {
            // print cell as char
            putchar(tape[ptr]);
        }
        else if (strcmp(trimmed, "loop[") == 0) {
            // start loop
            if (loop_stack_ptr < 1000) {
                loop_stack[loop_stack_ptr++] = i;
            }
        }
        else if (strcmp(trimmed, "]") == 0) {
            // end loop
            if (tape[ptr] != 0) {
                if (loop_stack_ptr > 0) {
                    i = loop_stack[loop_stack_ptr - 1];
                }
            } else {
                if (loop_stack_ptr > 0) {
                    loop_stack_ptr--;
                }
            }
        }
        else {
            printf(RED "error caught while parsing\n");
            printf(RED "at line: %s\n", trimmed);
            break;
        }
    }
    
    printf("\n");
    free(tape);
    free(code_copy);
}
